### PR TITLE
attributes migration

### PR DESCRIPTION
**Pull request by robkooper**
Add a new table that allows for attributes in json form #505
Add dump/load bety scripts from PEcAn to BETY #597

### CODE REVIEW

File: db/migrate/20181129000515_create_attributes_table.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+class CreateAttributesTable < ActiveRecord::Migration[5.1]
+  def change
+    this_hostid = Machine.new.hostid
+
+    create_table :attributes, id: :bigint do |t|
+      t.string :container_type, null: false
+      t.integer :container_id, limit: 8, null: false
+      t.jsonb :value, null: false, default: '{}'
```
Comment (gsrohde) [inline db/migrate/20181129000515_create_attributes_table.rb]: Would it make sense to constrain `value` in some way?  I assume this is just supposed to be a collection of key-value pairs.

File: db/migrate/20181129000515_create_attributes_table.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      t.timestamps
+    end
+
+    add_index :attributes, :container_id
+    add_index :attributes, :value, using: :gin
+
+    reversible do |dir|
+      dir.up do
+        execute %{
+          SELECT setval('attributes_id_seq', GREATEST(1, 1 + CAST(1e9 * #{this_hostid}::int AS bigint)), FALSE);
```
[the remark this answers is not in the document]
Comment (robkooper) [reply db/migrate/20181129000515_create_attributes_table.rb]: done

File: db/migrate/20181129000515_create_attributes_table.rb (hunk 1 above)
Comment (gsrohde) [reply db/migrate/20181129000515_create_attributes_table.rb]: Yes, but really _any_ JSON document?

Comment (robkooper) [reply db/migrate/20181129000515_create_attributes_table.rb]: yup, i don't want to limit this table. The whole idea of this table is for people to add any data to any existing row in the database.

Comment (gsrohde) [reply db/migrate/20181129000515_create_attributes_table.rb]: So even though we call this table "attributes", the json really could be anything, it doesn't even have to be a hash, yes?

Comment (robkooper) [reply db/migrate/20181129000515_create_attributes_table.rb]: it has to be json format that is all we know. We don't know what it will contain.

Comment (gsrohde) [reply db/migrate/20181129000515_create_attributes_table.rb]: OK, I'll just assume you know what you're doing.

**Comment by robkooper** [comment]
Waiting on merging this issue to see if we want to rush this before AGU

**Comment by gsrohde** [comment]
@dlebauer Rob addressed one minor coding nit that I flagged.  Regarding the larger concern having to do with lack of constraints on the `value` column, I don't think I really am in a position to weigh in on this question.  I'm not versed in the theory or philosophy of using relational databases in this way, and I wasn't in on any discussion of what the motive was for adding this table and how it would be used.  From what I could glean from the simple fact that this is called an _attributes_ table, I take it that this was seen as a method for adding "sparse" attributes to various database relations in an ad hoc way.  If this is the case, I would think there would at least be the requirement that the value be a JSON object (as opposed to an array or a primitive) with the keys serving as the attribute names and perhaps no restriction on the values.

I'm also guessing that this new table was motivated by needs coming from PEcAn and that the BETYdb web app will simply provide the migration and will otherwise ignore these changes.  Under this assumption, I'm going to go ahead and click Approve changes since I think the changes are largely out of the scope of both my expertise and what I am in charge of.

**Comment by gsrohde** [comment]
@robkooper @dlebauer Actually, I do have one concern not related to any previously expressed concerns—namely, that this requires at least PostgreSQL 9.4 and the official documentation only calls for version 9.3 or did until very recently.  (I somehow can access the technical documentation link, so I can't check what it says now.)  I think all the deployments I work with—EBI, bety-mepp, and terra-ref—now are using 9.4 or later, but I don't know if this is the case with all the other public instances.

**Comment by robkooper** [comment]
https://github.com/PecanProject/betydb-documentation/blob/master/installing_betydb/installing-betydb-web-application.md talks about PostgreSQL 9.3 or greater. other places in the documentation talks about postgresql 9.4

**Comment by gsrohde** [comment]
@robkooper Yes, I noticed that, when the site eventually came up for me.  I also made this change on the `bookdown` branch of the documentation, which will eventually replace the master branch when we deploy the bookdown version to some official place.  (It can't replace it now, because GitBook will choke on some of the bookdown syntax.)  If you think it make sense to, I'll rename the `bookdown` branch to `develop`.

**Pull request merged**